File: Python/ExampleTests/common/shipping_input_string.py
from . import tokens
DNC_STRING = '?DNC?'


class ShippingInputString:
    def __init__(self, total_price: str = '', shipping_cost: str = '', notes: str = ''):
        self.total_price = total_price
        self.shipping_cost = shipping_cost
        self.notes = notes

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 3, 'ShippingInput')
        return cls(
            parts[0],
            parts[1],
            parts[2]
        )

    def __str__(self):
        return (tokens.token(self.total_price) + ' ' +
                tokens.token(self.shipping_cost) + ' ' +
                tokens.token(self.notes))

    def _key(self):
        return (self.total_price, self.shipping_cost, self.notes)

    def __eq__(self, other):
        if not isinstance(other, ShippingInputString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'ShippingInputString({self})'

File: Swift/ExampleTests/common/ShippingInputString.swift
public struct ShippingInputString: CustomStringConvertible, Equatable {
    public let totalPrice: String
    public let shippingCost: String
    public let notes: String

    public init(totalPrice: String, shippingCost: String, notes: String) {
        self.totalPrice = totalPrice
        self.shippingCost = shippingCost
        self.notes = notes
    }

    public init(fromArray v: [String]) {
        self.totalPrice = v.count > 0 ? v[0] : ""
        self.shippingCost = v.count > 1 ? v[1] : ""
        self.notes = v.count > 2 ? v[2] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> ShippingInputString {
        let parts = Tokens.require(text, 3, "ShippingInput")
        return ShippingInputString(totalPrice: parts[0], shippingCost: parts[1], notes: parts[2])
    }

    public var description: String {
        return Tokens.token(totalPrice) + " " + Tokens.token(shippingCost) + " " + Tokens.token(notes)
    }

    public static let dncString = "?DNC?"

    public static func == (a: ShippingInputString, b: ShippingInputString) -> Bool {
        return (a.totalPrice == dncString || b.totalPrice == dncString || a.totalPrice == b.totalPrice)
            && (a.shippingCost == dncString || b.shippingCost == dncString || a.shippingCost == b.shippingCost)
            && (a.notes == dncString || b.notes == dncString || a.notes == b.notes)
    }
}
